File: src/engraft/actions/file_replace.py
import shutil
from dataclasses import dataclass
from pathlib import Path

from engraft.actions import register
from engraft.actions.base import Action


@register("file_replace")
@dataclass
class FileReplace(Action):
    """Replace a file in the working directory with a source file."""

    target: Path
    source_path: Path

    def apply(self) -> None:
        if not self.source_path.exists():
            raise FileNotFoundError(f"Source file does not exist: {self.source_path}")

        if not self.target.exists():
            raise FileNotFoundError(f"Target file does not exist: {self.target}")

        shutil.copy2(self.source_path, self.target)

File: typescript/src/actions/fileReplace.ts
import { promises as fs, existsSync } from "node:fs";
import path from "node:path";

import type { Action, Variables } from "./base.js";
import { register } from "./registry.js";

@register("file_replace")
export class FileReplace implements Action {
  private readonly file: string;
  private readonly variable: string;

  constructor(config: Record<string, unknown>) {
    this.file = String(config.file);
    this.variable = String(config.variable);
  }

  async apply(
    variables: Variables,
    workDir: string,
    valuesDir: string,
  ): Promise<void> {
    const value = variables[this.variable];
    if (value == null) return;

    const sourcePath = path.resolve(valuesDir, value);
    const targetPath = path.join(workDir, this.file);

    if (!existsSync(sourcePath)) {
      throw new Error(`Source file does not exist: ${sourcePath}`);
    }
    if (!existsSync(targetPath)) {
      throw new Error(`Target file does not exist: ${targetPath}`);
    }

    await fs.copyFile(sourcePath, targetPath);
  }

  targetFiles(): string[] {
    return [this.file];
  }
}
